pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -24,66 +24,268 @@ use crate::{
+    pub(super) async fn try_freeze_in_memory_layer(
+        &mut self,
+        lsn: Lsn,
+        last_freeze_at: &AtomicLsn,
+        write_lock: &mut tokio::sync::MutexGuard<'_, Option<TimelineWriterState>>,
+    ) -> Result<bool, Shutdown> {
+        Ok(self
+            .open_mut()?
+            .try_freeze_in_memory_layer(lsn, last_freeze_at, write_lock)
+            .await)
+    }
+
+    /// Add image layers to the layer map, called from `create_image_layers`.
+    pub(crate) fn track_new_image_layers(
+        &mut self,
+        image_layers: &[ResidentLayer],
+        metrics: &TimelineMetrics,
+    ) -> Result<(), Shutdown> {
+        self.open_mut()?
+            .track_new_image_layers(image_layers, metrics);
+        Ok(())
+    }
+
+    /// Flush a frozen layer and add the written delta layer to the layer map.
+    pub(crate) fn finish_flush_l0_layer(
+        &mut self,
+        delta_layer: Option<&ResidentLayer>,
+        frozen_layer_for_check: &Arc<InMemoryLayer>,
+        metrics: &TimelineMetrics,
+    ) -> Result<(), Shutdown> {
+        self.open_mut()?
+            .finish_flush_l0_layer(delta_layer, frozen_layer_for_check, metrics);
+        Ok(())
+    }
+
+    /// LayerManager shutdown. The in-memory layers do cleanup on drop, so we must drop them in
+    /// order to allow shutdown to complete.
+    ///
+    /// If there was a want to flush in-memory layers, it must have happened earlier.
+    pub(crate) fn shutdown(&mut self, writer_state: &mut Option<TimelineWriterState>) {
+        use LayerManager::*;
+        match self {
+            Open(OpenLayerManager {
+                layer_map,
+                layer_fmgr: LayerFileManager(hashmap),
+            }) => {
+                let open = layer_map.open_layer.take();
+                let frozen = layer_map.frozen_layers.len();
+                let taken_writer_state = writer_state.take();
+                tracing::info!(open = open.is_some(), frozen, "dropped inmemory layers");
+                let layers = std::mem::take(hashmap);
+                *self = Closed { layers };
+                assert_eq!(open.is_some(), taken_writer_state.is_some());
+            }
+            Closed { .. } => {
+                tracing::debug!("ignoring multiple shutdowns on layer manager")
+            }
+        }
+    }
+
+    /// Called when compaction is completed.
+    pub(crate) fn finish_compact_l0(
+        &mut self,
+        compact_from: &[Layer],
+        compact_to: &[ResidentLayer],
+        metrics: &TimelineMetrics,
+    ) -> Result<(), Shutdown> {
+        self.open_mut()?
+            .finish_compact_l0(compact_from, compact_to, metrics);
+        Ok(())
+    }
+
+    /// Called when a GC-compaction is completed.
+    pub(crate) fn finish_gc_compaction(
+        &mut self,
+        compact_from: &[Layer],
+        compact_to: &[ResidentLayer],
+        metrics: &TimelineMetrics,
+    ) -> Result<(), Shutdown> {
+        self.open_mut()?
```

Comment:
These wrappers feel unnecessary: callers already have to be aware of the concept that their LayerManager might not be initialized (i.e. handling `Shutdown`), so it isn't onerous to also make the caller do a `layers.open_mut()?.finish_gc_compaction` rather than maintaining this set of 2 line wrappers

Reply:
Yes, pushing this soon.

Reply:
077a9a4a4